--- turba-2.1.2/lib/Driver/ldap.php	2006-08-02 11:30:32.313829885 +0300
+++ co.php?r=1.83&p=1&f=turba%2Flib%2FDriver%2Fldap.php	2006-08-02 11:32:25.536404778 +0300
@@ -684,12 +660,16 @@
             return $schema;
         }
 
-        $attv = $schema->get('attribute', $att);
-        if (isset($attv['syntax'])) {
-            return $attv['syntax'];
-        } else {
-            return $this->_getSyntax($attv['sup'][0]);
+        if (!isset($this->_syntaxCache[$att])) {
+            $attv = $schema->get('attribute', $att);
+            if (isset($attv['syntax'])) {
+                $this->_syntaxCache[$att] = $attv['syntax'];
+            } else {
+                $this->_syntaxCache[$att] = $this->_getSyntax($attv['sup'][0]);
+            }
         }
+
+        return $this->_syntaxCache[$att];
     }
 
     /**
